tokio-stream/src/stream_ext/map_while.rs

```diff
@@ -0,0 +1,72 @@
+use crate::Stream;
+
+use core::fmt;
+use core::pin::Pin;
+use core::task::{Context, Poll};
+use pin_project_lite::pin_project;
+
+pin_project! {
+    /// Stream for the [`map_while`](super::StreamExt::map_while) method.
+    #[must_use = "streams do nothing unless polled"]
+    pub struct MapWhile<St, F> {
+        #[pin]
+        stream: St,
+        f: F,
+        done: bool
```

Comment:
This lets the `Option` optimization make the struct smaller.
```suggestion
        #[pin]
        stream: Option<St>,
        f: F,
```

Reply:
How do you know the option optimization applies here? Wouldn't that depend on the monomorphized `St` type? Is it common for the optimization to apply to `Stream` implementations?

Reply:
Yes, it depends on the underlying `Stream`, but it includes any stream containing pretty much any kind of pointer, and that's very common.

Reply:
Thanks!

A perhaps interesting aspect of this change is that the `St` object will be dropped as soon as the `f` returns `None`. I believe this is an acceptable aspect of this API, but should that be documented explicitly? The same does not apply to `take_while`, for example, where the `St` is not dropped until the `TakeWhile` is dropped, even if the `TakeWhile` predicate returns `false`.

Reply:
I don't think it's necessary to document that it happens.